### PR TITLE
Rework functions api with consumer

### PR DESCRIPTION
**Pull request by aahmed-se**
Functions api lacks ability to use a consumer interface
and the counter interface can be simplified.

### CODE REVIEW

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * This is one of the core interface of the function api. The process is called
+ * for every message of the input topic of the function. The incoming input bytes
+ * are converted to the input type I for simple Java types(String, Integer, Boolean,
+ * Map, and List types) and for org.Json type. If this serialization approach does not
+ * meet your needs, you can use the byte stream handler defined in RawRequestHandler.
+ */
+@FunctionalInterface
+public interface Consumer<I> {
```
Comment (srkukarni) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java]: we need to make this implement Function<I, Void>

Comment (aahmed-se) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java]: That's not right we need to add support at the functions runtime

Comment (srkukarni) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java]: That will be a far bigger change. Making this abstract class and implementing as Function might be the easiest way to go

Comment (aahmed-se) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java]: I don't think that's better I have adde changes to support the functions runtime

**Comment by aahmed-se** [comment]
The target java community will want popular function interfaces to be represented in our sdk, which is only a handful, I don't think we want to  restrict the sdk to a single interface.

**Comment by sijie** [comment]
since I prefer simplicity and consistency across languages, I am -0 on this change. however if other people feel this is needed, go for it.

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/Utils.java [hunk 2]
```diff
@@ -130,7 +131,12 @@ public static int findAvailablePort() {
             if (userClass instanceof Function) {
                 Function pulsarFunction = (Function) userClass;
                 typeArgs = TypeResolver.resolveRawArguments(Function.class, pulsarFunction.getClass());
-            } else {
+            }
+            else if (userClass instanceof Consumer){
```
Comment (jerrypeng) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/Utils.java]: Is this code tested?  I don't think this currently is even going to run.  Many places in the functions code we assume there are two types returned by the this function.

Comment (aahmed-se) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/Utils.java]: Not tested, if we can agree on the approach I will add the remaining code and tests.

**Comment by jerrypeng** [comment]
I would suggest add this only if there is actual user demand.  Writing a function "Function<Foo, Void>" vs "Consumer<Foo>" isn't that much different, but now we have to also to consider "Consumer" interface in java function runtime which will add complexity.

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java (hunk 1 above)
Comment (jerrypeng) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Consumer.java]: I rather the function runtime not have to worry about another interface

**Pull request closed without merging**